### ISSUE TITLE
Detecting misuse of reserved identifiers

### ISSUE THREAD

**Issue by pascal-cuoq** [TrustInSoft/tis-interpreter#87]
tis-interpreter working from the pre-processed source code, some misuses of [reserved identifiers](http://port70.net/~nsz/c/c11/n1570.html#7.1.3) are impossible to detect, but tis-interpreter should catch those that are visible after pre-processing.

Example:
https://asciinema.org/a/b28ilqd84qe77k5gsgahhv0y1 ([source](https://twitter.com/hasherezade/status/728841802988564480))

**Comment by ch3root** [comment]
There are many cases of such optimizations.

1. A libc call emitted by a compiler where there was none, without optimization:
  - initialization of aggregates -> memset;
  - assignment of structs -> memcpy.

2. A libc call emitted by a compiler where there was none, for optimization:
  - manual byte-by-byte setting -> memset;
  - manual byte-by-byte copying -> memcpy.

3. One libc call replaced by another one for optimization:
  - printf -> puts;
  - printf -> putchar;
  - malloc+memset -> calloc (gcc);
  - sprintf - > strcpy (gcc) / strlen+memcpy (clang);

  etc.

  gcc has a list of builtin function  -- https://gcc.gnu.org/onlinedocs/gcc/Other-Builtins.html . They can be optimized in some cases.

4. Interposed calls between libc functions.

  AIUI in glibc interposition only of malloc &co. is supported.

**Comment by ch3root** [comment]
Example with initialization of aggregates.

Source code:
```c
#include <stdio.h>

void *memset(void *p, int c, unsigned long z)
{
  (void)p, (void)c, (void)z;

  puts("Hi!");

  return 0;
}

int main()
{
  char s[10000] = "";
  (void)s;
}
```

tis-interpreter (e8de317f91c7250b3f83abc99783fe093dd88a29) output:
```
[value] Analyzing a complete application starting at main
[value] Computing initial state
[value] Initial state computed
[value] done for function main
```

gcc (GCC) 7.0.0 20160502 (experimental) w/ubsan output:
```
$ gcc -std=c11 -pedantic -Wall -Wextra -fsanitize=undefined test.c && ./a.out
Hi!
```

clang version 3.9.0 (trunk 268228) w/ubsan output:
```
$ clang -std=c11 -Weverything -fsanitize=undefined test.c && ./a.out
Hi!
```

**Comment by ch3root** [comment]
BTW names starting with two underscores etc. should be caught too.

Source code:
```c
#include <stdio.h>

int __builtin_putchar(int c);
int __builtin_putchar(int c)
{
  printf("Hi!\n");
  
  return c;
}

int main()
{
  printf("\n");
}
```

tis-interpreter (095c3a888e4af1d29f733688f2d47f0cc881cebd) output:
```
[value] Analyzing a complete application starting at main
[value] Computing initial state
[value] Initial state computed



[value] done for function main
```

gcc (GCC) 7.0.0 20160613 (experimental):
```
$ gcc -std=c11 -pedantic -Wall -Wextra -O3 -fsanitize=undefined test.c && ./a.out
Hi!
```

clang version 3.9.0 (trunk 271312):
```
$ clang -std=c11 -Weverything -O3 -fsanitize=undefined test.c && ./a.out

```

**Comment by pascal-cuoq** [comment]
Oh, I see. I was surprised that you said that just using a `__` prefix would suffice to solve a problem earlier in a discussion about the libc. It does as long as all uses of `__` identifiers are forbidden in user code.

This would require to distinguish system files from user files, which, after pre-processing, is a bit annoying to do, and cannot be done with complete accuracy before pre-processing. A brute-force solution is to pre-process with standard headers that do not contain a single `__` identifier, detect any occurrence of `__` in the resulting pre-processed files, throw away and pre-process again with the real headers, and continue as usual. No OCaml knowledge is required and this can be done in parallel with the interpretation. Sounds like a perfect student project.

**Comment by ch3root** [comment]
Yeah, identifiers beginning with _[A-Z_] are always reserved. This has two consequences:
- everything can break if a program tries to use them so tis-interpreter should catch it;
- tis-interpreter is free to use them for its own purposes.

I don't think such issues could be solved reliably without own preprocessor -- see #109. In particular, preprocessing with two different sets of headers is not the most reliable approach. But maybe it's good enough in practice. Another possible approach is to prepare a blacklist of reserved names. Not sure which way is better.
